Add NNC AOT Compiler executable

priyaramani opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #63994

Differential Revision: [D30582149](https://our.internmc.facebook.com/intern/diff/D30582149)


Goal: Add ability to do Ahead Of Time (AOT) compilation of models. 
This PR adds an API to compile models AOT . `aot_model_compiler` cpp executable will compile the forward method of the given model and produce LLVM assembly code and compiled model files as output.

On a bigger picture, the goal of AOT compilation is to be able to achieve extremely small pytorch binary size. 
As a bonus, we get better performance with a faster runtime too.

Thanks Mikhail and Jiakai for the prototype! 

Steps build aot_compiler and compile mobilenetv3 model:
```
# Build PyTorch with LLVM
USE_LLVM=<LLVM install path> python setup.py develop

# Get mobilenetv3 model
jf download GICWmACPJxxBk_YQAO8qqJoT-VkMbsIXAAAz --file mobilenetv3.pt

# Build aot_model_compiler cpp executable 
./scripts/build_local.sh -DBUILD_BINARY=ON

# Run aot_model_compiler on mobilenetv3 model
build/bin/aot_model_compiler --model mobilenetv3.pt --model_name=pytorch_dev_mobilenetv3 --model_version=v1 --input_dims="1,3,224,224"
```

Output:
```
Input sizes: 1, 3, 224, 224,
graph(%x.1 : Tensor):
  %2 : Float(1280, 1000, strides=[1, 1280], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
  %3 : int[] = prim::Constant[value=[2]]()
  %4 : int[] = prim::Constant[value=[3]]()
  %5 : Float(144, 576, strides=[1, 144], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
  %6 : Float(576, 144, strides=[1, 576], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
  %7 : Float(144, 576, strides=[1, 144], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
  %8 : Float(576, 144, strides=[1, 576], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
  %9 : Float(72, 288, strides=[1, 72], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
  %10 : Float(288, 72, strides=[1, 288], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
.
.
.
.
 %self.features.14.bias, %24, %23, %24, %29) # /home/mvz/pytorch/torch/nn/modules/conv.py:386:15
  %420 : Tensor(1, 1280, 1, 1) = aten::add(%input14.1, %26, %29) # mobilenetv3.py:31:27
  %421 : Tensor(1, 1280, 1, 1) = aten::hardtanh(%420, %28, %27) # /home/mvz/pytorch/torch/nn/functional.py:1161:17
  %422 : Tensor(1, 1280, 1, 1) = aten::mul(%input14.1, %421) # mobilenetv3.py:31:15
  %x0.1 : Tensor(1, 1280, 1, 1) = aten::div(%422, %27) # mobilenetv3.py:31:15
  %424 : Tensor(1, 1280, 1) = aten::mean(%x0.1, %4, %self.features.0.1.training, %self.features.0.0.bias) # mobilenetv3.py:205:12
  %x1.1 : Tensor(1, 1280) = aten::mean(%424, %3, %self.features.0.1.training, %self.features.0.0.bias) # mobilenetv3.py:205:12
  %ret0.1 : Tensor(1, 1000) = aten::addmm(%self.classifier.1.bias, %x1.1, %2, %29, %29) # /home/mvz/pytorch/torch/nn/functional.py:1665:14
  return (%ret0.1)

The compiled model was saved to mobilenetv3.compiled.pt
```

Outputs compiled model and LLVM assembly code files
```
(pytorch)  ~/local/pytorch aot
└─ $ ls | grep mobilenetv3.compiled
mobilenetv3.compiled.ll
mobilenetv3.compiled.pt
```

Further integration tests at [D30595123](https://our.internmc.facebook.com/intern/diff/D30595123)

priyaramani committed:
Add NNC AOT Compiler executable

[ghstack-poisoned]

priyaramani committed:
Update on "Add NNC AOT Compiler executable"




[ghstack-poisoned]

priyaramani committed:
Update on "Add NNC AOT Compiler executable"




[ghstack-poisoned]

priyaramani committed:
Update on "Add NNC AOT Compiler executable"




[ghstack-poisoned]

priyaramani committed:
Update on "Add NNC AOT Compiler executable"


Differential Revision: [D30582149](https://our.internmc.facebook.com/intern/diff/D30582149)

[ghstack-poisoned]

bertmaher commented on torch/csrc/jit/mobile/nnc/aot_compiler.cpp:
```diff
@@ -0,0 +1,120 @@
+    const std::string& method_name,
+    std::shared_ptr<Graph>& g,
+    const std::vector<int64_t>& sizes,
+    std::string* compiled_assembly) {
```
Minor: slight preference for returning a `pair` or a custom struct rather than a pointer out param.

bertmaher reviewed (approved):
Looks good, only minor comments from me.

ZolotukhinM commented on torch/csrc/jit/mobile/nnc/aot_compiler.cpp:
```diff
@@ -0,0 +1,120 @@
+// (c) Facebook, Inc. and its affiliates. Confidential and proprietary.
```
Is it confidential and proprietary though?

ZolotukhinM commented on torch/csrc/jit/mobile/nnc/aot_compiler.cpp:
```diff
@@ -0,0 +1,120 @@
+
+#include "aot_compiler.h"
+
+#include <ATen/Functions.h>
+#include <ATen/NativeFunctions.h>
+#include <torch/csrc/jit/ir/ir.h>
+#include <torch/csrc/jit/jit_opt_limit.h>
+#include <torch/csrc/jit/passes/constant_propagation.h>
+#include <torch/csrc/jit/passes/dead_code_elimination.h>
+#include <torch/csrc/jit/tensorexpr/ir.h>
+#include <torch/csrc/jit/tensorexpr/ir_simplifier.h>
+#include <torch/csrc/jit/tensorexpr/kernel.h>
+#include <torch/csrc/jit/tensorexpr/llvm_codegen.h>
+#include <torch/csrc/jit/tensorexpr/loopnest.h>
```
Some includes here are unnecessary:
```
ir_simplifier.h
loopnest.h
llvm_codegen.h
jit_opt_limit.h
```
It's a good practice only include what's used in the file. Also, please run clang-format to sort includes alphabetically.

ZolotukhinM commented on torch/csrc/jit/mobile/nnc/aot_compiler.cpp:
```diff
@@ -0,0 +1,120 @@
+  g = tensorexpr::removeUnusedSelfArgument(g);
+
+  std::vector<c10::optional<at::Tensor>> example_inputs = {at::rand(sizes)};
+  g->dump();
```
We probably don't need to dump the graph here. If you do want to keep debugging statements, please use `GRAPH_DUMP(g)` - this macro can be toggled on/off with `PYTORCH_JIT_LOG_LEVEL` environment variable (see more details on this here: https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/OVERVIEW.md#jit-logging).

ZolotukhinM commented on torch/csrc/jit/mobile/nnc/aot_compiler.cpp:
```diff
@@ -0,0 +1,120 @@
+    const std::vector<int64_t>& sizes,
+    std::string* compiled_assembly) {
+  auto g2 = g->copy();
+  std::cerr << "Input sizes: ";
+  for (auto s : sizes) {
+    std::cerr << s << ", ";
+  }
+  std::cerr << "\n";
```
Let's cleanup debug prints, or wrap them with `GRAPH_DEBUG` macros.

ZolotukhinM commented on torch/csrc/jit/mobile/nnc/aot_compiler.cpp:
```diff
@@ -0,0 +1,120 @@
+  PeepholeOptimize(g, false);
+  ConstantPropagation(g);
+  PropagateShapesOnGraph(g);
+  g->dump();
```
Same

ZolotukhinM commented on torch/csrc/jit/mobile/nnc/aot_compiler.h:
```diff
@@ -0,0 +1,23 @@
+namespace nnc {
+
+
+TORCH_API std::unique_ptr<Function> aot_compile(
```
Please add a brief comment for the function.

ZolotukhinM commented on torch/csrc/jit/mobile/nnc/aot_compiler.cpp:
```diff
@@ -0,0 +1,120 @@
+  return r;
+}
+
+void get_compiled_function(
```
Nit: even though we're not very consistent with this, I think our guidelines are to use lowerCamelCase for function names:
> * We follow the Facebook C++ Style Guidelines. Examples: some_namespace, SomeClass, SomeEnum, someFunction, someMethod, someMemberVariable, somePrivateMemberVariable_, SOME_MACRO
> * Exception: C++ methods and functions bound into python use python naming scheme (snake_case, ...) also in their C++ names. Exposing a previously unexposed method might need renaming the C++ method.

priyaramani replied:
Thanks for the review! 
Addressed the comments.

ZolotukhinM reviewed (approved):
👍

priyaramani committed:
Update on "Add NNC AOT Compiler executable"


Differential Revision: [D30582149](https://our.internmc.facebook.com/intern/diff/D30582149)


Goal: Add ability to do Ahead Of Time (AOT) compilation of models. 
On a bigger picture, the goal of AOT compilation is to be able to achieve extremely small pytorch binary size. 
As a bonus, we get better performance with a faster runtime too.

This PR adds an API to compile models AOT . `aot_model_compiler` cpp executable will compile the forward method of the given model and produce LLVM assembly code and compiled model files as output.

Thanks Mikhail and Jiakai for the original implementation and prototype! Refer:[D28902495](https://our.internmc.facebook.com/intern/diff/D28902495) [D28943489](https://our.internmc.facebook.com/intern/diff/D28943489) [D28089081](https://our.internmc.facebook.com/intern/diff/D28089081)

Steps build aot_compiler and compile mobilenetv3 model:
```
# Build PyTorch with LLVM

# Get model
jf download GICWmACPJxxBk_YQAO8qqJoT-VkMbsIXAAAz --file mobilenetv3.pt

# Build aot_model_compiler cpp executable 
./scripts/build_local.sh -DBUILD_BINARY=ON

# Run aot_model_compiler on mobilenetv3 model
build/bin/aot_model_compiler --model mobilenetv3.pt --model_name=pytorch_dev_mobilenetv3 --model_version=v1 --input_dims="1,3,224,224"
```

Output:
```
Input sizes: 1, 3, 224, 224,
graph(%x.1 : Tensor):
  %2 : Float(1280, 1000, strides=[1, 1280], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
  %3 : int[] = prim::Constant[value=[2]]()
  %4 : int[] = prim::Constant[value=[3]]()
  %5 : Float(144, 576, strides=[1, 144], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
  %6 : Float(576, 144, strides=[1, 576], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
  %7 : Float(144, 576, strides=[1, 144], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
  %8 : Float(576, 144, strides=[1, 576], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
  %9 : Float(72, 288, strides=[1, 72], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
  %10 : Float(288, 72, strides=[1, 288], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
.
.
.
.
 %self.features.14.bias, %24, %23, %24, %29) # /home/mvz/pytorch/torch/nn/modules/conv.py:386:15
  %420 : Tensor(1, 1280, 1, 1) = aten::add(%input14.1, %26, %29) # mobilenetv3.py:31:27
  %421 : Tensor(1, 1280, 1, 1) = aten::hardtanh(%420, %28, %27) # /home/mvz/pytorch/torch/nn/functional.py:1161:17
  %422 : Tensor(1, 1280, 1, 1) = aten::mul(%input14.1, %421) # mobilenetv3.py:31:15
  %x0.1 : Tensor(1, 1280, 1, 1) = aten::div(%422, %27) # mobilenetv3.py:31:15
  %424 : Tensor(1, 1280, 1) = aten::mean(%x0.1, %4, %self.features.0.1.training, %self.features.0.0.bias) # mobilenetv3.py:205:12
  %x1.1 : Tensor(1, 1280) = aten::mean(%424, %3, %self.features.0.1.training, %self.features.0.0.bias) # mobilenetv3.py:205:12
  %ret0.1 : Tensor(1, 1000) = aten::addmm(%self.classifier.1.bias, %x1.1, %2, %29, %29) # /home/mvz/pytorch/torch/nn/functional.py:1665:14
  return (%ret0.1)

The compiled model was saved to mobilenetv3.compiled.pt
```

Outputs compiled model and LLVM assembly code files
```
(pytorch)  ~/local/pytorch aot
└─ $ ls | grep mobilenetv3.compiled
mobilenetv3.compiled.ll
mobilenetv3.compiled.pt
```

Further integration tests at [D30595123](https://our.internmc.facebook.com/intern/diff/D30595123)

[ghstack-poisoned]

priyaramani commented:
@priyaramani has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30582149).

IvanKobzarev committed:
Update on "Add NNC AOT Compiler executable"


Differential Revision: [D30582149](https://our.internmc.facebook.com/intern/diff/D30582149)


Goal: Add ability to do Ahead Of Time (AOT) compilation of models. 
On a bigger picture, the goal of AOT compilation is to be able to achieve extremely small pytorch binary size. 
As a bonus, we get better performance with a faster runtime too.

This PR adds an API to compile models AOT . `aot_model_compiler` cpp executable will compile the forward method of the given model and produce LLVM assembly code and compiled model files as output.

Thanks Mikhail and Jiakai for the original implementation and prototype! Refer:[D28902495](https://our.internmc.facebook.com/intern/diff/D28902495) [D28943489](https://our.internmc.facebook.com/intern/diff/D28943489) [D28089081](https://our.internmc.facebook.com/intern/diff/D28089081)

Steps build aot_compiler and compile mobilenetv3 model:
```
# Build PyTorch with LLVM

# Get model
jf download GICWmACPJxxBk_YQAO8qqJoT-VkMbsIXAAAz --file mobilenetv3.pt

# Build aot_model_compiler cpp executable 
./scripts/build_local.sh -DBUILD_BINARY=ON

# Run aot_model_compiler on mobilenetv3 model
build/bin/aot_model_compiler --model mobilenetv3.pt --model_name=pytorch_dev_mobilenetv3 --model_version=v1 --input_dims="1,3,224,224"
```

Output:
```
Input sizes: 1, 3, 224, 224,
graph(%x.1 : Tensor):
  %2 : Float(1280, 1000, strides=[1, 1280], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
  %3 : int[] = prim::Constant[value=[2]]()
  %4 : int[] = prim::Constant[value=[3]]()
  %5 : Float(144, 576, strides=[1, 144], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
  %6 : Float(576, 144, strides=[1, 576], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
  %7 : Float(144, 576, strides=[1, 144], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
  %8 : Float(576, 144, strides=[1, 576], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
  %9 : Float(72, 288, strides=[1, 72], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
  %10 : Float(288, 72, strides=[1, 288], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
.
.
.
.
 %self.features.14.bias, %24, %23, %24, %29) # /home/mvz/pytorch/torch/nn/modules/conv.py:386:15
  %420 : Tensor(1, 1280, 1, 1) = aten::add(%input14.1, %26, %29) # mobilenetv3.py:31:27
  %421 : Tensor(1, 1280, 1, 1) = aten::hardtanh(%420, %28, %27) # /home/mvz/pytorch/torch/nn/functional.py:1161:17
  %422 : Tensor(1, 1280, 1, 1) = aten::mul(%input14.1, %421) # mobilenetv3.py:31:15
  %x0.1 : Tensor(1, 1280, 1, 1) = aten::div(%422, %27) # mobilenetv3.py:31:15
  %424 : Tensor(1, 1280, 1) = aten::mean(%x0.1, %4, %self.features.0.1.training, %self.features.0.0.bias) # mobilenetv3.py:205:12
  %x1.1 : Tensor(1, 1280) = aten::mean(%424, %3, %self.features.0.1.training, %self.features.0.0.bias) # mobilenetv3.py:205:12
  %ret0.1 : Tensor(1, 1000) = aten::addmm(%self.classifier.1.bias, %x1.1, %2, %29, %29) # /home/mvz/pytorch/torch/nn/functional.py:1665:14
  return (%ret0.1)

The compiled model was saved to mobilenetv3.compiled.pt
```

Outputs compiled model and LLVM assembly code files
```
(pytorch)  ~/local/pytorch aot
└─ $ ls | grep mobilenetv3.compiled
mobilenetv3.compiled.ll
mobilenetv3.compiled.pt
```

Further integration tests at [D30595123](https://our.internmc.facebook.com/intern/diff/D30595123)

[ghstack-poisoned]

priyaramani commented on binaries/aot_model_compiler.cc:
```diff
@@ -0,0 +1,172 @@
+    "semicolon to separate the dimension of different "
+    "tensors.");
+C10_DEFINE_string(
+    output_asm,
```
Thanks for the review!
Addressed the comments.

bertmaher commented:
@IvanKobzarev it looks like somehow reverted this PR to an earlier version?

IvanKobzarev commented:
@ivankobzarev has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30582149).

IvanKobzarev commented:
@ivankobzarev has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30582149).

priyaramani commented:
> @IvanKobzarev it looks like somehow reverted this PR to an earlier version?

priyaramani committed:
Update on "Add NNC AOT Compiler executable"


Differential Revision: [D30582149](https://our.internmc.facebook.com/intern/diff/D30582149)


Goal: Add ability to do Ahead Of Time (AOT) compilation of models. 
On a bigger picture, the goal of AOT compilation is to be able to achieve extremely small pytorch binary size. 
As a bonus, we get better performance with a faster runtime too.

This PR adds an API to compile models AOT . `aot_model_compiler` cpp executable will compile the forward method of the given model and produce LLVM assembly code and compiled model files as output.

Thanks Mikhail and Jiakai for the original implementation and prototype! Refer:[D28902495](https://our.internmc.facebook.com/intern/diff/D28902495) [D28943489](https://our.internmc.facebook.com/intern/diff/D28943489) [D28089081](https://our.internmc.facebook.com/intern/diff/D28089081)

Steps build aot_compiler and compile mobilenetv3 model:
```
# Build PyTorch with LLVM

# Get model
jf download GICWmACPJxxBk_YQAO8qqJoT-VkMbsIXAAAz --file mobilenetv3.pt

# Build aot_model_compiler cpp executable 
./scripts/build_local.sh -DBUILD_BINARY=ON

# Run aot_model_compiler on mobilenetv3 model
build/bin/aot_model_compiler --model mobilenetv3.pt --model_name=pytorch_dev_mobilenetv3 --model_version=v1 --input_dims="1,3,224,224"
```

Output:
```
Input sizes: 1, 3, 224, 224,
graph(%x.1 : Tensor):
  %2 : Float(1280, 1000, strides=[1, 1280], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
  %3 : int[] = prim::Constant[value=[2]]()
  %4 : int[] = prim::Constant[value=[3]]()
  %5 : Float(144, 576, strides=[1, 144], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
  %6 : Float(576, 144, strides=[1, 576], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
  %7 : Float(144, 576, strides=[1, 144], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
  %8 : Float(576, 144, strides=[1, 576], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
  %9 : Float(72, 288, strides=[1, 72], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
  %10 : Float(288, 72, strides=[1, 288], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
.
.
.
.
 %self.features.14.bias, %24, %23, %24, %29) # /home/mvz/pytorch/torch/nn/modules/conv.py:386:15
  %420 : Tensor(1, 1280, 1, 1) = aten::add(%input14.1, %26, %29) # mobilenetv3.py:31:27
  %421 : Tensor(1, 1280, 1, 1) = aten::hardtanh(%420, %28, %27) # /home/mvz/pytorch/torch/nn/functional.py:1161:17
  %422 : Tensor(1, 1280, 1, 1) = aten::mul(%input14.1, %421) # mobilenetv3.py:31:15
  %x0.1 : Tensor(1, 1280, 1, 1) = aten::div(%422, %27) # mobilenetv3.py:31:15
  %424 : Tensor(1, 1280, 1) = aten::mean(%x0.1, %4, %self.features.0.1.training, %self.features.0.0.bias) # mobilenetv3.py:205:12
  %x1.1 : Tensor(1, 1280) = aten::mean(%424, %3, %self.features.0.1.training, %self.features.0.0.bias) # mobilenetv3.py:205:12
  %ret0.1 : Tensor(1, 1000) = aten::addmm(%self.classifier.1.bias, %x1.1, %2, %29, %29) # /home/mvz/pytorch/torch/nn/functional.py:1665:14
  return (%ret0.1)

The compiled model was saved to mobilenetv3.compiled.pt
```

Outputs compiled model and LLVM assembly code files
```
(pytorch)  ~/local/pytorch aot
└─ $ ls | grep mobilenetv3.compiled
mobilenetv3.compiled.ll
mobilenetv3.compiled.pt
```

Further integration tests at [D30595123](https://our.internmc.facebook.com/intern/diff/D30595123)

[ghstack-poisoned]

priyaramani commented:
@priyaramani has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30582149).

priyaramani committed:
Update on "Add NNC AOT Compiler executable"


Differential Revision: [D30582149](https://our.internmc.facebook.com/intern/diff/D30582149)


Goal: Add ability to do Ahead Of Time (AOT) compilation of models. 
On a bigger picture, the goal of AOT compilation is to be able to achieve extremely small pytorch binary size. 
As a bonus, we get better performance with a faster runtime too.

This PR adds an API to compile models AOT . `aot_model_compiler` cpp executable will compile the forward method of the given model and produce LLVM assembly code and compiled model files as output.

Thanks Mikhail and Jiakai for the original implementation and prototype! Refer:[D28902495](https://our.internmc.facebook.com/intern/diff/D28902495) [D28943489](https://our.internmc.facebook.com/intern/diff/D28943489) [D28089081](https://our.internmc.facebook.com/intern/diff/D28089081)

Steps build aot_compiler and compile mobilenetv3 model:
```
# Build PyTorch with LLVM

# Get model
jf download GICWmACPJxxBk_YQAO8qqJoT-VkMbsIXAAAz --file mobilenetv3.pt

# Build aot_model_compiler cpp executable 
./scripts/build_local.sh -DBUILD_BINARY=ON

# Run aot_model_compiler on mobilenetv3 model
build/bin/aot_model_compiler --model mobilenetv3.pt --model_name=pytorch_dev_mobilenetv3 --model_version=v1 --input_dims="1,3,224,224"
```

Output:
```
Input sizes: 1, 3, 224, 224,
graph(%x.1 : Tensor):
  %2 : Float(1280, 1000, strides=[1, 1280], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
  %3 : int[] = prim::Constant[value=[2]]()
  %4 : int[] = prim::Constant[value=[3]]()
  %5 : Float(144, 576, strides=[1, 144], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
  %6 : Float(576, 144, strides=[1, 576], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
  %7 : Float(144, 576, strides=[1, 144], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
  %8 : Float(576, 144, strides=[1, 576], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
  %9 : Float(72, 288, strides=[1, 72], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
  %10 : Float(288, 72, strides=[1, 288], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
.
.
.
.
 %self.features.14.bias, %24, %23, %24, %29) # /home/mvz/pytorch/torch/nn/modules/conv.py:386:15
  %420 : Tensor(1, 1280, 1, 1) = aten::add(%input14.1, %26, %29) # mobilenetv3.py:31:27
  %421 : Tensor(1, 1280, 1, 1) = aten::hardtanh(%420, %28, %27) # /home/mvz/pytorch/torch/nn/functional.py:1161:17
  %422 : Tensor(1, 1280, 1, 1) = aten::mul(%input14.1, %421) # mobilenetv3.py:31:15
  %x0.1 : Tensor(1, 1280, 1, 1) = aten::div(%422, %27) # mobilenetv3.py:31:15
  %424 : Tensor(1, 1280, 1) = aten::mean(%x0.1, %4, %self.features.0.1.training, %self.features.0.0.bias) # mobilenetv3.py:205:12
  %x1.1 : Tensor(1, 1280) = aten::mean(%424, %3, %self.features.0.1.training, %self.features.0.0.bias) # mobilenetv3.py:205:12
  %ret0.1 : Tensor(1, 1000) = aten::addmm(%self.classifier.1.bias, %x1.1, %2, %29, %29) # /home/mvz/pytorch/torch/nn/functional.py:1665:14
  return (%ret0.1)

The compiled model was saved to mobilenetv3.compiled.pt
```

Outputs compiled model and LLVM assembly code files
```
(pytorch)  ~/local/pytorch aot
└─ $ ls | grep mobilenetv3.compiled
mobilenetv3.compiled.ll
mobilenetv3.compiled.pt
```

Further integration tests at [D30595123](https://our.internmc.facebook.com/intern/diff/D30595123)

[ghstack-poisoned]

priyaramani commented:
@priyaramani has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30582149).

priyaramani commented:
@priyaramani has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30582149).

priyaramani committed:
Update on "Add NNC AOT Compiler executable"


Differential Revision: [D30582149](https://our.internmc.facebook.com/intern/diff/D30582149)


Goal: Add ability to do Ahead Of Time (AOT) compilation of models. 
On a bigger picture, the goal of AOT compilation is to be able to achieve extremely small pytorch binary size. 
As a bonus, we get better performance with a faster runtime too.

This PR adds an API to compile models AOT . `aot_model_compiler` cpp executable will compile the forward method of the given model and produce LLVM assembly code and compiled model files as output.

Thanks Mikhail and Jiakai for the original implementation and prototype! Refer:[D28902495](https://our.internmc.facebook.com/intern/diff/D28902495) [D28943489](https://our.internmc.facebook.com/intern/diff/D28943489) [D28089081](https://our.internmc.facebook.com/intern/diff/D28089081)

Steps build aot_compiler and compile mobilenetv3 model:
```
# Build PyTorch with LLVM

# Get model
jf download GICWmACPJxxBk_YQAO8qqJoT-VkMbsIXAAAz --file mobilenetv3.pt

# Build aot_model_compiler cpp executable 
./scripts/build_local.sh -DBUILD_BINARY=ON

# Run aot_model_compiler on mobilenetv3 model
build/bin/aot_model_compiler --model mobilenetv3.pt --model_name=pytorch_dev_mobilenetv3 --model_version=v1 --input_dims="1,3,224,224"
```

Output:
```
Input sizes: 1, 3, 224, 224,
graph(%x.1 : Tensor):
  %2 : Float(1280, 1000, strides=[1, 1280], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
  %3 : int[] = prim::Constant[value=[2]]()
  %4 : int[] = prim::Constant[value=[3]]()
  %5 : Float(144, 576, strides=[1, 144], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
  %6 : Float(576, 144, strides=[1, 576], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
  %7 : Float(144, 576, strides=[1, 144], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
  %8 : Float(576, 144, strides=[1, 576], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
  %9 : Float(72, 288, strides=[1, 72], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
  %10 : Float(288, 72, strides=[1, 288], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
.
.
.
.
 %self.features.14.bias, %24, %23, %24, %29) # /home/mvz/pytorch/torch/nn/modules/conv.py:386:15
  %420 : Tensor(1, 1280, 1, 1) = aten::add(%input14.1, %26, %29) # mobilenetv3.py:31:27
  %421 : Tensor(1, 1280, 1, 1) = aten::hardtanh(%420, %28, %27) # /home/mvz/pytorch/torch/nn/functional.py:1161:17
  %422 : Tensor(1, 1280, 1, 1) = aten::mul(%input14.1, %421) # mobilenetv3.py:31:15
  %x0.1 : Tensor(1, 1280, 1, 1) = aten::div(%422, %27) # mobilenetv3.py:31:15
  %424 : Tensor(1, 1280, 1) = aten::mean(%x0.1, %4, %self.features.0.1.training, %self.features.0.0.bias) # mobilenetv3.py:205:12
  %x1.1 : Tensor(1, 1280) = aten::mean(%424, %3, %self.features.0.1.training, %self.features.0.0.bias) # mobilenetv3.py:205:12
  %ret0.1 : Tensor(1, 1000) = aten::addmm(%self.classifier.1.bias, %x1.1, %2, %29, %29) # /home/mvz/pytorch/torch/nn/functional.py:1665:14
  return (%ret0.1)

The compiled model was saved to mobilenetv3.compiled.pt
```

Outputs compiled model and LLVM assembly code files
```
(pytorch)  ~/local/pytorch aot
└─ $ ls | grep mobilenetv3.compiled
mobilenetv3.compiled.ll
mobilenetv3.compiled.pt
```

Further integration tests at [D30595123](https://our.internmc.facebook.com/intern/diff/D30595123)

[ghstack-poisoned]

priyaramani commented:
@priyaramani has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30582149).

priyaramani commented:
@priyaramani has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30582149).

priyaramani committed:
Update on "Add NNC AOT Compiler executable"


Differential Revision: [D30582149](https://our.internmc.facebook.com/intern/diff/D30582149)


Goal: Add ability to do Ahead Of Time (AOT) compilation of models. 
On a bigger picture, the goal of AOT compilation is to be able to achieve extremely small pytorch binary size. 
As a bonus, we get better performance with a faster runtime too.

This PR adds an API to compile models AOT . `aot_model_compiler` cpp executable will compile the forward method of the given model and produce LLVM assembly code and compiled model files as output.

Thanks Mikhail and Jiakai for the original implementation and prototype! Refer:[D28902495](https://our.internmc.facebook.com/intern/diff/D28902495) [D28943489](https://our.internmc.facebook.com/intern/diff/D28943489) [D28089081](https://our.internmc.facebook.com/intern/diff/D28089081)

Steps build aot_compiler and compile mobilenetv3 model:
```
# Build PyTorch with LLVM
USE_LLVM=<LLVM install path> python setup.py develop

# Get model
jf download GICWmACPJxxBk_YQAO8qqJoT-VkMbsIXAAAz --file mobilenetv3.pt

# Build aot_model_compiler cpp executable 
./scripts/build_local.sh -DBUILD_BINARY=ON

# Run aot_model_compiler on mobilenetv3 model
build/bin/aot_model_compiler --model mobilenetv3.pt --model_name=pytorch_dev_mobilenetv3 --model_version=v1 --input_dims="1,3,224,224"
```

Output:
```
Input sizes: 1, 3, 224, 224,
graph(%x.1 : Tensor):
  %2 : Float(1280, 1000, strides=[1, 1280], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
  %3 : int[] = prim::Constant[value=[2]]()
  %4 : int[] = prim::Constant[value=[3]]()
  %5 : Float(144, 576, strides=[1, 144], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
  %6 : Float(576, 144, strides=[1, 576], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
  %7 : Float(144, 576, strides=[1, 144], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
  %8 : Float(576, 144, strides=[1, 576], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
  %9 : Float(72, 288, strides=[1, 72], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
  %10 : Float(288, 72, strides=[1, 288], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
.
.
.
.
 %self.features.14.bias, %24, %23, %24, %29) # /home/mvz/pytorch/torch/nn/modules/conv.py:386:15
  %420 : Tensor(1, 1280, 1, 1) = aten::add(%input14.1, %26, %29) # mobilenetv3.py:31:27
  %421 : Tensor(1, 1280, 1, 1) = aten::hardtanh(%420, %28, %27) # /home/mvz/pytorch/torch/nn/functional.py:1161:17
  %422 : Tensor(1, 1280, 1, 1) = aten::mul(%input14.1, %421) # mobilenetv3.py:31:15
  %x0.1 : Tensor(1, 1280, 1, 1) = aten::div(%422, %27) # mobilenetv3.py:31:15
  %424 : Tensor(1, 1280, 1) = aten::mean(%x0.1, %4, %self.features.0.1.training, %self.features.0.0.bias) # mobilenetv3.py:205:12
  %x1.1 : Tensor(1, 1280) = aten::mean(%424, %3, %self.features.0.1.training, %self.features.0.0.bias) # mobilenetv3.py:205:12
  %ret0.1 : Tensor(1, 1000) = aten::addmm(%self.classifier.1.bias, %x1.1, %2, %29, %29) # /home/mvz/pytorch/torch/nn/functional.py:1665:14
  return (%ret0.1)

The compiled model was saved to mobilenetv3.compiled.pt
```

Outputs compiled model and LLVM assembly code files
```
(pytorch)  ~/local/pytorch aot
└─ $ ls | grep mobilenetv3.compiled
mobilenetv3.compiled.ll
mobilenetv3.compiled.pt
```

Further integration tests at [D30595123](https://our.internmc.facebook.com/intern/diff/D30595123)

[ghstack-poisoned]

priyaramani commented:
@priyaramani has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30582149).
